espresso/environment/query_service_intercept.go

```diff
@@ -0,0 +1,387 @@
+	return DecisionProxy
+}
+
+// proxyRequest is a simple HTTP handler that proxies requests to the given
+// baseURL, utilizing the given http.Client.
+type proxyRequest struct {
+	client  *http.Client
+	baseURL url.URL
+}
+
+// ServeHTTP implements http.Handler
+func (p *proxyRequest) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	buf := new(bytes.Buffer)
+	if _, err := io.Copy(buf, r.Body); err != nil && err != io.EOF {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	req, err := http.NewRequest(r.Method, p.baseURL.JoinPath(r.URL.Path).String(), buf)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Copy over the headers
+	for k, v := range r.Header {
+		req.Header.Set(k, v[0])
+	}
+
+	res, err := p.client.Do(req)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer res.Body.Close()
+
+	buf.Reset()
+	if _, err := io.Copy(buf, res.Body); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(res.StatusCode)
+	for k, v := range res.Header {
+		w.Header().Set(k, v[0])
+	}
+
+	// Write the proxy response contents
+	if _, err := io.Copy(w, buf); err != nil {
+		// If we encounter an error here, it will be difficult to actually
+		// handle it at this point, as we've already sent the response headers.
+		//
+		// The best we can do at this point, is log the error.
+		_ = err
+		return
+	}
+}
+
+// fakeSubmitTransactionSuccess is a simple HTTP handler that simulates a
+// successful transaction submission by returning a fake commit hash.
+type fakeSubmitTransactionSuccess struct{}
+
+// generateCommitForSubmitTransaction generates a commit hash for the
+// transaction in the request body. This is a fake implementation that
+// simulates a successful transaction submission by returning a commit hash
+// that won't collide with the real transaction commit hashes.
+func generateCommitForSubmitTransaction(r *http.Request) (*types.TaggedBase64, error) {
+	defer r.Body.Close()
+
+	var txn types.Transaction
+	if err := json.NewDecoder(r.Body).Decode(&txn); err != nil {
+		// Unable to decode, this is a problem?
+		var emptyHash [32]byte
+		return tagged_base64.New("FAKE", emptyHash[:])
+	}
+
+	commit := txn.Commit()
+	return tagged_base64.New("FAKE", commit[:])
+}
+
+// ServeHTTP implements http.Handler
+func (fakeSubmitTransactionSuccess) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// We could do a lot of effort to validate the request, and return an
+	// hash that is actually representative of the transaction that was
+	// just submitted. In some cases we may actually want this sort of
+	// validated behavior, but it's very simple to just return any hash
+	// instead.
+
+	// We should probably validate the request contents and format here, but
+	// we will just assume the settings.
+	defer r.Body.Close()
```

Comment:
Should not we also simulate a query that just fails in addition to returning a randomly generated hash value?

Reply:
We **COULD** also do this.  The acceptance criteria for the test stated to turn the `builder` on and off randomly, which would not result in any noticeable affect via the query endpoint.  Though, the original criteria also stated to turn off random espresso nodes, which **could** exhibit this kind of behavior.

Reply:
Would it be a lot of work?

Reply:
I could simulate a load balancer returning a status code indicating that the service is unreachable.  I won't be able to simulate an error connecting to the service itself.  That would require intercepting the http client itself, which we don't have the ability to do with the current `espresso-network-go` library.

Reply:
> I could simulate a load balancer returning a status code indicating that the service is unreachable. I won't be able to simulate an error connecting to the service itself. 

Why can't the fake http server return error some status code from time to time (instead of return a random hash)? Or is this what you mean?

> That would require intercepting the http client itself, which we don't have the ability to do with the current espresso-network-go library.

Right no need to do that I think.

Reply:
> Why can't the fake http server return some error status code from time to time (instead of return a random hash)? Or is this what you mean?

This is exactly what I mean.  I would return a `503` to simulate the service being unavailable. It doesn't really matter which status code is returned as long as it is an error status code.  I just figured Service Unavailable would be the easiest thing to concretely reason about as opposed to something like [I'm a teapot](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/418).

Reply:
Oh! And other status codes, specifically those in the `4xx` range might have an assumed serialized response body of `application/json` which would be more effort to represent effectively.

Reply:
Done in ee150d3c745734b16d2b880c74f38d940fd6cce8.

Reply:
That was fast!